Sources/AsyncAlgorithms/MultiProducerSingleConsumerChannel/MultiProducerSingleConsumerChannel.swift

```diff
@@ -0,0 +1,489 @@
+///
+/// Values can be send to the source from asynchronous contexts using ``MultiProducerSingleConsumerChannel/Source/send(_:)-9b5do``
+/// and ``MultiProducerSingleConsumerChannel/Source/send(contentsOf:)-4myrz``. Backpressure results in calls
+/// to the `send` methods to be suspended. Once more elements should be produced the `send` methods will be resumed.
+///
+/// ```
+/// try await withThrowingTaskGroup(of: Void.self) { group in
+///     group.addTask {
+///         try await source.send(1)
+///         try await source.send(2)
+///         try await source.send(3)
+///     }
+///
+///     for await element in channel {
+///         print(element)
+///     }
+/// }
+/// ```
+///
+/// ### Synchronous producers
+///
+/// Values can also be send to the source from synchronous context. Backpressure is also exposed on the synchronous contexts; however,
+/// it is up to the caller to decide how to properly translate the backpressure to underlying producer e.g. by blocking the thread.
+///
+/// ## Finishing the source
+///
+/// To properly notify the consumer if the production of values has been finished the source's ``MultiProducerSingleConsumerChannel/Source/finish(throwing:)`` **must** be called.
+public struct MultiProducerSingleConsumerChannel<Element, Failure: Error>: AsyncSequence {
+    /// A private class to give the ``MultiProducerSingleConsumerChannel`` a deinit so we
+    /// can tell the producer when any potential consumer went away.
+    private final class _Backing: Sendable {
+        /// The underlying storage.
+        fileprivate let storage: _Storage
+
+        init(storage: _Storage) {
+            self.storage = storage
+        }
+
+        deinit {
+            storage.sequenceDeinitialized()
+        }
+    }
+
+    /// The backing storage.
+    private let backing: _Backing
+
+    @frozen
+    public struct ChannelAndStream: ~Copyable {
+        public var channel: MultiProducerSingleConsumerChannel
+        public var source: Source
+
+        public init(
+            channel: MultiProducerSingleConsumerChannel,
+            source: consuming Source
+        ) {
+            self.channel = channel
+            self.source = source
+        }
+    }
+
+    /// Initializes a new ``MultiProducerSingleConsumerChannel`` and an ``MultiProducerSingleConsumerChannel/Source``.
+    ///
+    /// - Parameters:
+    ///   - elementType: The element type of the channel.
+    ///   - failureType: The failure type of the channel.
+    ///   - BackpressureStrategy: The backpressure strategy that the channel should use.
+    /// - Returns: A tuple containing the channel and its source. The source should be passed to the
+    ///   producer while the channel should be passed to the consumer.
+    public static func makeChannel(
+        of elementType: Element.Type = Element.self,
+        throwing failureType: Failure.Type = Never.self,
+        backpressureStrategy: Source.BackpressureStrategy
+    ) -> ChannelAndStream {
```

Comment:
What's the reason for this returning a nominal type? Previously it returned a tuple which allows for initialization of the sequence as follows:

```swift
let (channel, source) = MultiProducerSingleConsumerChannel.makeChannel(/* ... */)
```

This is also what the documentation comment on the type suggests.

However, with the nominal type you cannot do that and, if you are storing these as properties, are forced to do this...

```swift
let channelAndSource = MultiProducerSingleConsumerChannel.makeChannel(/* ... */)
self.channel = channelAndSource.channel
self.source = channelAndSource.source
```

...instead of this...

```swift
(self.channel, self.source) = MultiProducerSingleConsumerChannel.makeChannel(/* ... */)
```

Reply:
Sadly, tuples don't support `~Copyable` types. That's the only reason for a nominal type here.